SS1: Support arbitrary sublist names.

paulclarkaranz opened the pull request:
Adds support for using record sublists other than `item` and `addressbook`. This shouldn't require any changes in dependant pieces of
code unless they were relying on matching an error message or on `selectNewLineItem` erroring on unsupported records (it also "upserts" in the SS2 module).

I have already raised a PR upstream with this change (by mistake...) but thought I should do one here to or else that would mean pushing to master without review!

Most places that would have previously complained about unsupported lines will now complain about `TypeError: Cannot * property '*' of undefined`. Exceptions are:
* `selectNewLineItem`: Will create new sublist groups and new currentLine groups as appropriate in line with what `selectNewLine` does in the SS2 module. (I'm not sure what that `addressbookaddress` thing is for but left it in for compatibility.)
* `getLineItemCount` it was already returning 0 for unknown sublists so I left it that way

I have left `lineItems` and `addressBookLines` as aliases in case they are being used in existing tests somehow (I don't think that is possible given the scope though). `lineItems` is still used in `transform()`, I am not sure if that should be updated and if so what to.

paulclarkaranz committed:
SS1: Support arbitrary sublist names.

Adds support for using record sublists other than `item` and
`addressbook`. This shouldn't require any changes in dependant pieces of
code unless they were relying on matching an error message or on
`selectNewLineItem` erroring on unsupported records (it also "upserts"
in the SS2 module).

Most places that would have previously complained about unsupported lines will
now complain about `TypeError: Cannot * property '*' of undefined`. Exceptions
are:
* `selectNewLineItem`: Will create new sublist groups and new currentLine
  groups as appropriate in line with what `selectNewLine` does in the SS2
  module. (I'm not sure what that `addressbookaddress` thing is for but left
  it in for compatibility.)
* `getLineItemCount` it was already returning 0 for unknown sublists so I left
  it that way

I have left `lineItems` and `addressBookLines` as aliases in case they are
being used in existing tests somehow (I don't think that is possible given the
scope though). `lineItems` is still used in `transform()`, I am not sure if
that should be updated and if so what to.

paulclarkaranz committed:
SS1: Add setCurrentLineItemText

Similar to in 3EN-Cloud/netsumo#12, add `setCurrentLineItemText` to set
line item list fields by display string.

And the same caveat applies, getting the value later won't return the
index like it would in NetSuite.

jscn commented on modules/SS1/nlobjRecord.js:
```diff
@@ -39,62 +43,39 @@ var nlobjRecord = function (recordtype, internalid) {
+    }
+    currentLineItems[group] = {}
+    currentLineItems[group]['id'] = id+'_'+getLineItemCount(group);
+    currentLineItems[group]['line'] = getLineItemCount(group);
```
Any particular reason you're constructing this object one property at a time, rather than just using an object literal?

jscn reviewed (approved):
Looks reasonable to me. Will you try and get this merged upstream?

paulclarkaranz committed:
SS1: Return search operator result from matchesRecord.

Previously the search `nlobjSearchFilter.matchesRecord()` was always
returning undefined so `nlapiSearchRecord` should never have been
matching anything.

paulclarkaranz committed:
SS1: Adds `nlobjSearchResult.getText()`.

In NetSuite this returns the display string for select fields. Here it
returns the value, like everything else.
